supporting-blog-content/evaluating-search-relevance-part-1/retrieve-and-rerank.ipynb

```diff
@@ -0,0 +1,808 @@
+    "    doc_ids = [doc_id for doc_id, _ in bm25_res.items()]\n",
+    "    if len(doc_ids) == 0:\n",
+    "        results_after_reranking[qid] = dict()\n",
+    "        continue\n",
+    "\n",
+    "    doc_texts = [corpus_dict[doc_id] for doc_id in doc_ids]\n",
+    "\n",
+    "    # rescore with the reranking model\n",
+    "    scores = reranking_model.predict([(query_text, doc_text) for doc_text in doc_texts])\n",
+    "\n",
+    "    results_after_reranking[qid] = {\n",
+    "        doc_id: float(score) for doc_id, score in zip(doc_ids, scores)\n",
+    "    }"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a85e0f51-fedc-4490-800c-48fd268d8db7",
+   "metadata": {},
+   "source": [
+    "and let's calculate the metric scores for the reranked results"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1c145552-a372-4624-92f8-4798636dd3ec",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "post_reranking_eval_scores_per_query = evaluator.evaluate(results_after_reranking)\n",
+    "\n",
+    "post_reranking_eval_scores = defaultdict(list)\n",
+    "\n",
+    "for qid, vals in post_reranking_eval_scores_per_query.items():\n",
+    "    for metric, metric_score in vals.items():\n",
+    "        post_reranking_eval_scores[metric].append(metric_score)\n",
+    "\n",
+    "for metric, scores in post_reranking_eval_scores.items():\n",
+    "    print(f\"{metric}: {np.mean(scores)}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d35c3a11-9d85-4b5b-9c68-0add866f3700",
+   "metadata": {},
+   "source": [
+    "which in most cases will provide a significant boost in performance"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a5f50faf-9192-43b6-be35-6700b740881d",
+   "metadata": {},
+   "source": [
+    "## Bonus section\n",
+    "\n",
+    "Let's do some more extra analysis and try to answer the question `\"How many times is an evaluator presented with (query, document) pairs for which there is no ground truth information?\"`\n",
+    "In other words, we calculate the percentage of cases where the `qrels` file contains a relevance score for a particular document in the result list.\n",
+    "Let's start with BM25 by focusing on the **top-10** retrieved documents"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "d5fac2ad-3780-4931-9ed2-127a804fb9f3",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "TOP_K = 10\n",
+    "\n",
+    "judge_rate_per_query = []\n",
+    "\n",
+    "for qid, doc_scores in bm25_responses.items():\n",
+    "    top_k_doc_ids = [\n",
+    "        doc_id\n",
+    "        for doc_id, score in sorted(\n",
+    "            doc_scores.items(), key=lambda x: x[1], reverse=True\n",
+    "        )[:TOP_K]\n",
+    "    ]\n",
+    "    if len(top_k_doc_ids) == 0:\n",
+    "        continue\n",
+    "\n",
+    "    labeled_docs = [doc_id for doc_id in top_k_doc_ids if doc_id in qrels[qid]]\n",
```

Comment:
Style nit (no need to materialise the list):
```suggestion
    "    labeled_docs = sum(1 for _ in top_k_doc_ids if doc_id in qrels[qid])\n",
```

Reply:
Thanks, I need to adapt the line just below to make it compatible.